### PR TITLE
Make the RA Phase Tank uncloak during repair

### LINKED ISSUE #18106: STNK doesnt uncloak on Service Depot
Both in TD and RA mods, STNK lacks Heal given at UncloakOn, so it remains cloaked while being on a FIX, creating an awkward situation where from the rival player's point of view can see something is being repaired on the service depot but not what it actually is. On the other hand this is an non-issue as those are the only cloakable units that can be put on a FIX. Still, it might be good for consistency sake for them to uncloak while being repaired on a Service Depot.

### PR DESCRIPTION
**Pull request by abcdefg30**
Closes #18106.

### CODE REVIEW

**Comment by pchote** [comment]
Does using `Heal` here mean that they will recloak again if you run out of money during repair?

IMO service depots should trigger `Dock`, then we can use `Dock` on TD and `Dock | Heal` on RA (for mechanics).

**Comment by sorcerer86pt** [comment]
> Does using `Heal` here mean that they will recloak again if you run out of money during repair?
> 
> IMO service depots should trigger `Dock`, then we can use `Dock` on TD and `Dock | Heal` on RA (for mechanics).

I support this approach. Looks cleaner and having that trait should simplify code for other things (like in RA2, removing the terror drone when a unit is with Heal trait)

**Comment by abcdefg30** [comment]
Updated as suggested. This still has the problem of the actor not uncloaking when sent to repair on a depot while the player has no cash. But that is more of a problem of when `INotifyResupply.BeforeResupply` is called which means we either take this as is with this minor polish issue and wait until docking is fixed, or we rewrite parts of `Resupply` again.

**Review by Mailaender — DISMISSED** [review]
Looks fine.

**Comment by pchote** [comment]
~The PR title says TD and RA, but this only fixes RA!~

File: OpenRA.Mods.Common/Traits/Cloak.cs [hunk 1]
```diff
@@ -247,5 +248,16 @@ void INotifyInfiltration.Infiltrating(Actor self)
 			if (Info.UncloakOn.HasFlag(UncloakType.Infiltrate))
 				Uncloak();
 		}
+
+		void INotifyResupply.BeforeResupply(Actor host, Actor target, ResupplyType types)
+		{
+			if (Info.UncloakOn.HasFlag(UncloakType.Dock))
+			{
+				isDocking = true;
```
Comment (pchote) [inline OpenRA.Mods.Common/Traits/Cloak.cs:256]: I don't see a matching `isDocking = false`. How does this reset?

**Comment by pchote** [comment]
On closer investigation, this actually doesn't work at all! `INotifyResupply` is called on the service depot, not on the unit being resupplied!

The simplest fix may be to add a new `INotifyBeingResupplied` and then move the uncloaking into that interface's `ResupplyTick`. This would fix the issues with units not uncloaking when first entering a depot or recloaking while on it if the player runs out of cash. It will also remove the need to set `isDocking`.

**Comment by abcdefg30** [comment]
I see, this didn't come up since we also define `Heal` in RA. Not changing TD was a deliberate choice after talking to people on Discord.

**Comment by abcdefg30** [comment]
Updated.

**Review by pchote — APPROVED** [review]

**Review by teinarss — APPROVED** [review]

**Comment by teinarss** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/f69147af358e710cc7a0f8c01b561fdd98ade12f...c50e70f802ae6d94a83aad7bfca28c1f6cf25ab3

**Pull request merged**